File: src/main/java/br/unifor/costify/domain/valueobject/Money.java
package br.unifor.costify.domain.valueobject;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;

public final class Money {
    private static final int SCALE = 2; // 2 decimal places for currency
    private static final RoundingMode ROUNDING_MODE = RoundingMode.HALF_UP;
    
    private final BigDecimal amount;

    private Money(BigDecimal amount) {
        if (amount == null) {
            throw new IllegalArgumentException("Amount cannot be null");
        }
        if (amount.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        this.amount = amount.setScale(SCALE, ROUNDING_MODE);
    }

    public static Money of(BigDecimal amount) {
        return new Money(amount);
    }

    public static Money of(double amount) {
        return new Money(BigDecimal.valueOf(amount));
    }

    public static Money zero() {
        return new Money(BigDecimal.ZERO);
    }

    public Money add(Money other) {
        if (other == null) {
            throw new IllegalArgumentException("Money to add cannot be null");
        }
        return new Money(this.amount.add(other.amount));
    }

    public Money subtract(Money other) {
        if (other == null) {
            throw new IllegalArgumentException("Money to subtract cannot be null");
        }
        BigDecimal result = this.amount.subtract(other.amount);
        if (result.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("Result cannot be negative");
        }
        return new Money(result);
    }

    public Money multiply(double multiplier) {
        return multiply(BigDecimal.valueOf(multiplier));
    }

    public Money multiply(BigDecimal multiplier) {
        if (multiplier == null) {
            throw new IllegalArgumentException("Multiplier cannot be null");
        }
        return new Money(this.amount.multiply(multiplier));
    }

    public Money divide(double divisor) {
        return divide(BigDecimal.valueOf(divisor));
    }

    public Money divide(BigDecimal divisor) {
        if (divisor == null) {
            throw new IllegalArgumentException("Divisor cannot be null");
        }
        if (divisor.compareTo(BigDecimal.ZERO) == 0) {
            throw new IllegalArgumentException("Cannot divide by zero");
        }
        return new Money(this.amount.divide(divisor, SCALE, ROUNDING_MODE));
    }

    public boolean isGreaterThan(Money other) {
        if (other == null) {
            throw new IllegalArgumentException("Money to compare cannot be null");
        }
        return this.amount.compareTo(other.amount) > 0;
    }

    public boolean isLessThan(Money other) {
        if (other == null) {
            throw new IllegalArgumentException("Money to compare cannot be null");
        }
        return this.amount.compareTo(other.amount) < 0;
    }

    public boolean equals(Money other) {
        if (other == null) {
            return false;
        }
        return this.amount.compareTo(other.amount) == 0;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    public double doubleValue() {
        return amount.doubleValue();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Money money = (Money) obj;
        return Objects.equals(amount, money.amount);
    }

    @Override
    public int hashCode() {
        return Objects.hash(amount);
    }

    @Override
    public String toString() {
        return String.format("$%s", amount.toString());
    }
}

File: internal/domain/valueobject/money.go
package valueobject

import (
	"fmt"
	"math"

	"github.com/vini/costify-go/internal/domain/errors"
)

// Money represents a monetary value with business rules
type Money struct {
	amount float64
}

const (
	// Scale for monetary calculations (2 decimal places)
	MoneyScale = 100.0
)

// NewMoney creates a Money instance from a float64 amount
func NewMoney(amount float64) (Money, error) {
	if amount < 0 {
		return Money{}, errors.NewNegativeMoneyError("Money cannot be negative in business context")
	}
	// Round to 2 decimal places
	rounded := math.Round(amount*MoneyScale) / MoneyScale
	return Money{amount: rounded}, nil
}

// ZeroMoney creates a Money instance with zero value
func ZeroMoney() Money {
	return Money{amount: 0.0}
}

// Add adds two Money values
func (m Money) Add(other Money) Money {
	result := m.amount + other.amount
	rounded := math.Round(result*MoneyScale) / MoneyScale
	return Money{amount: rounded}
}

// Subtract subtracts another Money value from this one
func (m Money) Subtract(other Money) (Money, error) {
	result := m.amount - other.amount
	if result < 0 {
		return Money{}, errors.NewNegativeMoneyError("Subtraction would result in negative money")
	}
	rounded := math.Round(result*MoneyScale) / MoneyScale
	return Money{amount: rounded}, nil
}

// Multiply multiplies Money by a multiplier
func (m Money) Multiply(multiplier float64) (Money, error) {
	result := m.amount * multiplier
	if result < 0 {
		return Money{}, errors.NewNegativeMoneyError("Multiplication would result in negative money")
	}
	rounded := math.Round(result*MoneyScale) / MoneyScale
	return Money{amount: rounded}, nil
}

// Divide divides Money by a divisor
func (m Money) Divide(divisor float64) (Money, error) {
	if divisor == 0 {
		return Money{}, errors.NewDivisionByZeroError("Cannot divide money by zero")
	}
	result := m.amount / divisor
	if result < 0 {
		return Money{}, errors.NewNegativeMoneyError("Division would result in negative money")
	}
	rounded := math.Round(result*MoneyScale) / MoneyScale
	return Money{amount: rounded}, nil
}

// IsGreaterThan checks if this Money is greater than another
func (m Money) IsGreaterThan(other Money) bool {
	return m.amount > other.amount
}

// IsLessThan checks if this Money is less than another
func (m Money) IsLessThan(other Money) bool {
	return m.amount < other.amount
}

// Equals checks if two Money values are equal
func (m Money) Equals(other Money) bool {
	return math.Abs(m.amount-other.amount) < 0.001 // Handle floating point precision
}

// Amount returns the monetary amount as float64
func (m Money) Amount() float64 {
	return m.amount
}

// String implements the Stringer interface
func (m Money) String() string {
	return fmt.Sprintf("$%.2f", m.amount)
}
